Safe installation guide (#534)
safe-global/safe-docs: pages/core-api/safe-installation-guide.mdx

import { Callout, Steps, Tabs } from 'nextra/components'

# Safe stack installation guide

## Safe\{Core\} contracts deployment

> In the first step, you will deploy the Safe\{Core\} contracts on your chain.

Reviewer: ```suggestion
In this section, you will deploy the Safe\{Core\} contracts on your chain.
```

All Safe contract deployments on any network follow the same procedure to ensure a deterministic address for all singleton contracts (proxy-factory, mastercopy, etc.) and verify the deployment.

### Prerequisites

> Open a [pull request](https://github.com/ethereum-lists/chains) to add your chain to [chainlist.org](https://chainlist.org/).

Reviewer: Can this please be added to vale allowlist? Re: vale error.

</Callout>

1. Create a new issue in the [safe-singleton-factory](https://github.com/safe-global/safe-singleton-factory/issues/new?assignees=&labels=&projects=&template=new_chain.yml&title=%5BNew+chain%5D%3A+) repository.
2. A bot will reply to the issue with the deployer address (`0x914...3d7`) and the amount of native token you need to send to this address.
3. Once funded, mark the checkbox on the GitHub issue.
4. The review of the issues happens every two weeks. Our team will perform the deterministic deployment of the `safe-singleton-factory` contract and publish a new npm release of [@safe-global/safe-singleton-factory](https://www.npmjs.com/package/@safe-global/safe-singleton-factory).

### Singleton contracts deployment 

<Callout type="info" emoji="ℹ️">
 You do not need to perform these tasks if your network is based on a rollup framework with Safe contracts already deployed (for example, OP Stack).
</Callout>

1. Clone the [safe-smart-account](https://github.com/safe-global/safe-smart-account) repository by running the following command:
    ```

>     git clone --branch v1.3.0-libs.0 https://github.com/safe-global/safe-smart-account.git

Reviewer: We should add a cd command here as well.

Author: Agree.

    ```
    
2. Get the latest version of [@safe-global/safe-singleton-factory](https://www.npmjs.com/package/@safe-global/safe-singleton-factory), by running the following command:

    ```bash
    npm i --save-dev @safe-global/safe-singleton-factory
    ```

>     Ensure the latest version includes your `safe-singleton-factory` deployment from the earlier section.

Reviewer: Let's add a (relative) link to this section: #singleton-factory-contract-deployment

    ```json
    "<your_chain_id>": "canonical"
    ```

3. Open a pull request. Your pull request should follow this [example pull request](https://github.com/safe-global/safe-deployments/pull/679).

</Steps>

## Safe\{Core\} Infrastructure deployment

Once you have deployed the Safe\{Core\} contracts, you must deploy the off-chain components of the Safe\{Core\} Stack, including backend and frontend services. 

You have three options for deploying the Safe stack, depending on your needs.

### Option 1: Safe Core Contributor Platform-as-a-Service

We roll out new networks quarterly, depending on our internal capacity, and only for medium to large chains based on a strict scoring framework. 

If you want to submit your chain for assessment in the next quarter, please fill out this [form](https://noteforms.com/forms/request-safe-ui-and-infra-support-4weugt).

#### Hard requirements

- EVM-compatible chain.
- At least two dedicated RPC node providers, with preference for at least one Tier 1 provider of Infura, Alchemy, Quicknode. You can find more on RPC nodes [here](https://docs.safe.global/core-api/api-safe-transaction-service/rpc-requirements).

> ### Option 2: Third-party integrators

Reviewer: We should remove this section if we don't have data for this atm.

Author: I agree, but I will keep it there until after the call with Greg, if it is ok with you. Maybe they already have data.

Author: For now there will be an email link.

We are happy to introduce you to our official integrators.

List of official integrators:

    - integrator A | website | telegram | twitter
    - integrator B | website | telegram | twitter
    - integrator C | website | telegram | twitter

### Option 3: Self-hosting

> If you have engineering and Cloud resources, you can also deploy the different components of the Stack by following one of the procedures below. 

Reviewer: ```suggestion
If you have engineering and cloud resources, you can also deploy the different components of the stack by following one of the procedures below. 
```

> You can find docker images for a majority of platforms. 

Reviewer: Where can these be found? Can we link?

Author: I already asked the same question to the network team. :)

Author: Added.

#### Requirements

> 

Reviewer: We are missing the images in the page.

**Hardware**

These are the hardware requirements to run the different services required.

| Network | Transaction Service | Client Gateway | Config Service | Database |
| --- | --- | --- | --- | --- |
| Small (new chain with low traffic) | CPU: 2 vCPU RAM: 8GiB | CPU: 2 vCPU RAM: 8 GiB | CPU: 2 vCPU RAM: 8 GiB | CPU: 1 cores/ 2 vCPU RAM: 8GiB |
| Standard | CPU: 4 vCPU RAM: 16GiB | CPU: 2 vCPU RAM: 8 GiB | CPU: 2 vCPU RAM: 8 GiB | CPU: 2 cores/ 4 vCPU RAM: 16GiB |
| Large (large chain with high-traffic) | CPU: 8 vCPU RAM: 32GiB | CPU: 2 vCPU RAM: 8 GiB | CPU: 2 vCPU RAM: 8 GiB | CPU: 8 cores/ 16 vCPU RAM: 64GiB |

**RPC**

> For support, please refer to the [RPC requirements](https://docs.safe.global/core-api/api-safe-transaction-service/rpc-requirements#what-are-the-rpc-requirements). Safe (L1) requires tracing for indexing, while Safe (L2) supports events/logs.

Reviewer: Let's use relative links for the docs pages.

Author: Good point!

#### Docker-compose deployment

> Please see the [safe-infrastructure](https://github.com/safe-global/safe-infrastructure) repository for all required information.

Reviewer: Does that mean the section below does not contain the full details? If not, can we please add them here instead of pointing to that repo? We should also then remove this info from that repo and add a link to this page instead when released (to have a single source of truth).

Author: I discussed this with @gjeanmart; all required information is in this guide. I'm linking him here so he can decide if there is redundant information in the repo that could be removed. I hope that is fine for everyone.

```bash
git clone git@github.com:safe-global/safe-infrastructure.git
cd safe-infrastructure
```

**Prerequisites**

> - Docker v20.10+

Reviewer: Can we link to places to download these tools/software?

Author: Good idea, added the links.

- Docker-compose 2.x.x+
- Running Ethereum JSON RPC client

**Configuration**

```bash
cp .env.sample .env
vi .env
REVERSE_PROXY_PORT=8000
CFG_VERSION=latest
CGW_VERSION=latest
TXS_VERSION=latest
UI_VERSION=latest
EVENTS_VERSION=latest
RPC_NODE_URL=<REPLACE BY YOUR RPC ENDPOINT>
```

**Run**

> This command runs seventeen Docker containers from the [docker-compose.yml](https://github.com/gjeanmart/safe-infrastructure/blob/main/docker-compose.yml):

Reviewer: Can we link to a repo in the Safe org please?

- Nginx reverse proxy
- Postgres 14.x database (x3 for Transaction Service, Config Service, and Events Service)
- Redis database (x2 for Transaction Service and Client Gateway)
- RabbitMQ message queue (x2 for Transaction Service and General)
- Transaction Service (txs) workers & scheduler (x4)
- Transaction Service (txs) web

> - Config Service (cfg) web

Reviewer: Can the items in the parathesis cfg, ui, txs, etc can be removed?

Author: I guess so. Will remove and ask Greg for feedback.

- Client Gateway (cgx) web
- Events service (events) web
- Wallet (ui) web

```bash
sh scripts/run_locally.sh
# will ask to set up username/password for config-service and transactions-service
```

**Local Ganache node**

> You can optionally use a local Ganache node (including Safe's singleton contracts) for testing purposes:

Reviewer: We should link to Ganache.

```bash
vi .env
REVERSE_PROXY_PORT=8000
CFG_VERSION=latest
CGW_VERSION=latest
TXS_VERSION=latest
UI_VERSION=latest
EVENTS_VERSION=latest
CHAIN_ID=1337 # we recommend using 1337 to ensure compatibility with official deployments
RPC_NODE_URL=http://node:8545
MNEMONIC="test test test test test test test test test test test junk" # used to deploy singleton contracts

sh ./scripts/run_locally.sh -f docker-compose.local.yml
```

#### Kubernetes deployment

An official helm chart creates all the necessary manifests, including the service account and RBAC entities needed for service discovery.

```bash
helm repo add safe https://safe/safe/safe
helm repo update
helm install <RELEASE_NAME> safe/stack -f values.yaml [-n <NAMESPACE>]
```

The helm chart allows you to inline all the configurations directly in your `values.yaml`:

```bash
TO BE COMPLETED
```

> #### Troubleshoot

Reviewer: ```suggestion
#### Troubleshooting
```
